README
wmmount - The WindowMaker universal mount point
05/09/98  Release 1.0 Beta1
Copyright (C) 1998  Sam Hawker <shawkie@geocities.com>
This software comes with ABSOLUTELY NO WARRANTY
This software is free software, and you are welcome to redistribute it
under certain conditions
See the COPYING file for details.

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


THE AUTHOR:
===========

Sam Hawker (17/m/uk)

shawkie@geocities.com
http://www.geocities.com/SiliconValley/Vista/2471/


INSTALLING:
===========

xmkmf
make
strip wmmount
make install

mkdir /usr/X11R6/lib/X11/wmmount
cp EXTRAS/* /usr/X11R6/lib/X11/wmmount/


CONFIGURATION:
==============

The system-wide configuration file can now be generated automagically,
from /etc/fstab, by means of the system.wmmount.awk awk (or gawk) program.

awk -f system.wmmount.awk /etc/fstab >system.wmmount
cp system.wmmount /usr/X11R6/lib/X11/wmmount/

In addition, individual users can create their own configuration files,
saving them as ~/.wmmount


NOTE:
=====

I do not recommend giving wmmount suid or sgid permissions.
In fact, I strongly recommend not giving it such permissions.


USAGE:
======

Press the "<" and ">" arrow buttons to select a mount point.
Press the mount button (the one with a bolt symbol) to mount or unmount.
Double-click on the information box to open a program at the mount point
(Midnight Commander for example).

What else can i say? Enjoy!


AFTERSTEP USERS:
================
(thanks to tygris@erols.com)

This release includes in a new command line option.
The effect this has is to enable shape support (it implies "-s"), and
to reduce the size of the window to 56x56 pixels.
The "-position position" option has also been added, so you can push
wmcdplay off the edge of the screen while it gets swallowed.
To put wmmount in your Wharf, add the following line in the appropriate
part of your .steprc

*Wharf wmmount nil MaxSwallow "wmmount" wmmount -a -position -0-0 &

It is also possible (by editing and recompiling the afterstep sources),
to make Wharf handle the new 56x56 pixel window properly - yes, I add
an option to use a 56x56 pixel window, even though the AfterStep Wharf
really wants 55x57 pixel ones.

I am told (by tygris@erols.com):

"Locate Wharf.c (or is it Wharf.cc?)  Should be in
AfterStep-1.0/modules/Wharf.  Locate this:

            if (Buttons[button].maxsize) {
              Buttons[button].icons[0].w = 55;
              Buttons[button].icons[0].h = 57;
            }

and change the 55 and 57 to 56's.  Save and compile."


BUILD PROBLEMS:
===============

You may have trouble linking wmmount with the compiler supplied on some
recent distributions (eg. Debian 2.0 & RedHat 5.1).
It can be made to build by adding the following line to the top of the
Imakefile, before executing xmkmf:

CC = c++

I am not sure of the effect this has on memory usage, etc.


RELEASE HISTORY:
================

Release 0.1    25/01/98    Does not update shape transparency properly under
                              WindowMaker. This is a fault with WindowMaker.

Release 0.2    31/01/98    Now free's ALL pixmaps on exit (I think).

Release 0.3    04/02/98    Free's GC's too now.
                           Displays % capacity on mounted devices.
                           New set of icons (use a common palette of about 10 colors).
                           New format for ~/.wmmount.

Release 0.4    24/02/98    Fixed bug relating to manual window positioning.

Release 0.5    04/03/98    Updates capacity display correctly.
                              (thanks to aseltine@cs.umass.edu)
                           Reduced CPU Load.
                              (thanks to aseltine@cs.umass.edu)

Release 0.6    12/03/98    New option in .wmmount makes 'showcapacity' specific to
                              each mount.
                              (thanks to steeve@minet.net)
                           Now works on DEC Unix systems.
                              (thanks to aseltine@cs.umass.edu)
                           Now gives same capacity values as df.
                              (thanks to steeve@minet.net)
                           Now notices unmounts promptly (don't remove the floppy disk
                              just coz wmmount says its unmounted already!).
                           Now works on 8bpp displays with full colormaps.
                              (thanks to devernay@istar.fr on the mailing-list)
                           New fiend look.

Release 0.7    25/04/98    Compiles again now.
                           NFS and loopback mounts should work again now.
                           Now GPL.
                           Removed DEC Unix code.
                              (sorry, aseltine@cs.umass.edu)
                           Sensible CPU load.
                           New FAQ.
                           Yet another new face.
                              (thanks to warp@xs4all.nl)
                           Minor changes to ~/.wmmount.

Release 0.8    07/05/98    Removed broken display code.
                           Now looks for system.wmmount as a last resort.
                              (thanks to ssadiqr@pheonix.Princeton.EDU)
                           Added "-b back_color" command line argument.
                           Should work on FreeBSD now.
                              (thanks to dalroi@wit401310.student.utwente.nl)
                           Fixed permissions problem with NFS mounts.
                           Now uses execv(3) instead of system(3).

Release 1.0    05/09/98    Improved error messaging.
Beta1                      Fixed checkMount problem.
                              (thanks to unit@lgcy.com and others)
                           Less strict mountpoints, eg. "/mnt/dosc/" is now valid.
                           Added "-a" command line argument for AfterStep users.
                           Added "-position position" command line argument.
                           Command line arguments, "-a", "-w" and "-s" are now toggle,
                              so if you enable one at compile-time, you can override it
                              at run-time.
                           Added awk program for automatic generation of system.wmmount.
                              (thanks to rolfwr@ii.uib.no)
                           Added more options for device usage display.
                           Executes open command from the mount point directory (useful
                              for xterms, filemanagers, etc.).


KNOWN BUGS:
===========

Does not read symlinks properly when testing for mounted filesystems.
So, if it incorrectly reports devices as mounted, make sure the mountpoint
specified in the config file does not include symlinks.

debian/changelog
wmmount (1.0beta2-7.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS by adding xutils-dev to Build-Depends (Closes: #485271), as
    confirmed by Olly Betts. And remove xutils from there, as it is
    obsolete.
  * Set urgency to “medium” as the bug affects testing.
  * Menu transition: “Apps/System” to “Applications/System/Monitoring”.

 -- Cyril Brulebois <kibi@debian.org>  Sun, 19 Jul 2009 02:36:43 +0200

wmmount (1.0beta2-7) unstable; urgency=medium

  * Documented %% in the manual page, fully closes: #274875.
  * Check if XLoadQueryFont() returned NULL and if so, die nicely,
    rather than falling apart, closes: #392502.
  * Use execvp() instead of execv() for the opencmd command,
    thanks to Pierre Pronchery, closes: #300907.
  * Dropped references to /usr/X11R6 in the example config file.
  * Stopped shipping system.wmmount.sh in examples because we already have
    gen-wmmountrc which does the same.
  * Stopped shipping upstream README because it only has misleading info
    and nothing useful to users that isn't in the manual page already.
  * Replaced xterm with x-terminal-emulator in the default opencmd value
    because we only use -e and -T anyway.
  * Removed needless x-dev from build-depends.
  * Updated copyright file (URL, dropped pointless FSF address which dates).
  * Quoted the needs parameter in the menu file.
  * Updated Standards-Version (to 3.7.2).

 -- Josip Rodin <joy-packages@debian.org>  Sat, 25 Nov 2006 11:27:39 +0100

wmmount (1.0beta2-6.1) unstable; urgency=low

  * Non-maintainer upload to help xlibs-dev transition.
  * debian/control: Updated to replace xlibs-dev build-dep. (Closes: #346968)

 -- Marc 'HE' Brockschmidt <he@debian.org>  Fri, 20 Jan 2006 00:07:12 +0100

wmmount (1.0beta2-6) unstable; urgency=low

  * New maintainer address.

 -- Josip Rodin <joy-packages@debian.org>  Tue, 28 Jan 2003 15:26:11 +0100

wmmount (1.0beta2-5) unstable; urgency=low

  * Added build-depends on xutils for xmkmf, closes: #176541, #176547.

 -- Josip Rodin <jrodin@jagor.srce.hr>  Thu, 16 Jan 2003 01:31:04 +0100

wmmount (1.0beta2-4) unstable; urgency=low

  * Winter cleaning. :)
  * Fixed a spelling error in the description, closes: #125497.
    Rewrote the description.
  * Replaced Build-Depends: xlib6g-dev, libxpm4-dev with just xlibs-dev,
    closes: #170134.
    I guess this will screw me over when xlibs get split, but oh well.

 -- Josip Rodin <jrodin@jagor.srce.hr>  Sat, 11 Jan 2003 15:29:12 +0100

wmmount (1.0beta2-3) unstable; urgency=low

  * Killed the `User config file not found' message to stderr, it's
    pointless.
  * Changed menu entry to be in Apps/System, although that's not quite
    perfect either, but it closes: #59654.
  * Small changes to the manual page.
  * Policy 3.2.1 compliant:
    + checks for DEB_BUILD_OPTIONS in debian/rules
    + moved wmmount and its manual page out of /usr/X11R6 (I kept the .1x
      suffix for the manual page although it's in /usr/share/man/man1, it
      won't hurt anything)
  * Updated Build-Depends:.
  * Cleaned up copyright file.

 -- Josip Rodin <jrodin@jagor.srce.hr>  Sat,  9 Sep 2000 22:54:11 +0200

wmmount (1.0beta2-2) unstable; urgency=low

  * s/generate-wmmountrc/gen-wmmountrc/ in postinst. Closes: #51410.
  * Compiled with libstdc++2.10. Closes: #46968.
  * Standards-Version: 3.1.1. Added Build-Depends:.

 -- Josip Rodin <jrodin@jagor.srce.hr>  Sun, 28 Nov 1999 17:14:30 +0100

wmmount (1.0beta2-1) unstable; urgency=low

  * New upstream version. The config file syntax and command line
    arguments handling changed a lot, will produce errors and segfaults.
    So, a big fat warning is now in the postinst.
  * Included upstream config/system.wmmount.sh script in examples dir,
    and modified it to match file locations on Debian.
  * Updated update-wmmountrc to make it generate a new style .wmmount
    file. Renamed it to 'gen-wmmountrc', because it doesn't update any
    existing data, it just gen-erates new data, from /etc/fstab.
  * Updated upstream URL and wmmount.1x manual page. Moved the manpage
    from debian/wmmount.1x to wmmount.man, to be included in Makefile
    by default, and called 'install.man' rule, too.

 -- Josip Rodin <jrodin@jagor.srce.hr>  Sun, 17 Oct 1999 23:30:33 +0200

wmmount (1.0beta1-2) unstable; urgency=low

  * New maintainer.
  * Mostly redid the packaging, updated for Policy 3.x.

 -- Josip Rodin <jrodin@jagor.srce.hr>  Mon, 27 Sep 1999 17:59:09 +0200

wmmount (1.0beta1-1) unstable; urgency=low

  * New upstream version
  * debian/wmmount.1x: Updated switches and syntax
  * Configuration files reverted to 0.8's names
  * debian/wmmount.postinst: Modified to notify about configuration
    filename change
  * debian/control: Modified "Suggests" to "wmaker | afterstep"; updated
    Standards Version
  * debian/rules: added CC=c++ after $(MAKE)

 -- Marcelo E. Magallon <mmagallo@debian.org>  Sun,  4 Oct 1998 21:25:23 -0600

wmmount (0.8.3.unoff-1) unstable; urgency=low

  * New release by Steve Borho
  * Added preinst script to handle system wide config filename change
  * Modified postinst to notify of filename change
  * Added "ignore" to the man page

 -- Marcelo E. Magallon <mmagallo@debian.org>  Sun, 20 Sep 1998 12:01:57 -0600

wmmount (0.8.1.unoff-1) unstable; urgency=low

  * New upstream version by new author (Steve Borho <borho@stlnet.com>),
    ergo the "unoff" part.
  * Modified wmmount.c to move system wide file to /etc

 -- Marcelo E. Magallon <mmagallo@debian.org>  Sun,  6 Sep 1998 15:56:53 -0600

wmmount (0.8-1) unstable; urgency=low

  * New upstream version
  * Patched to let it properly determine mounted/unmounted filesystems
    (sent upstream)
  * Updated manpage

 -- Marcelo E. Magallon <mmagallo@debian.org>  Sat,  9 May 1998 21:52:45 -0600

wmmount (0.7-1) unstable; urgency=low

  * New upstream version
  * It's GPL'ed now, moving into main!
  * Applied patch that handles Expose events


wmmount.c
/* wmmount - The Window Maker universal mount point
 *
 * 17/06/99  Release 1.0 beta2
 * Copyright (C) 1999  Sam Hawker <shawkie@geocities.com>
 * This software comes with ABSOLUTELY NO WARRANTY
 * This software is free software, and you are welcome to redistribute it
 * under certain conditions
 * See the README file for details.
 */


/* User defines
 */

/* Intervals
 *   INTERVAL  - Minimum interval unit in usecs
 *   INTERVAL1 - Update interval in multiples of def_interval
 *   INTERVAL2 - Button repeat interval in multiples of def_interval
 *   INTERVAL3 - Double-click interval in multiples of def_interval
 */
#define INTERVAL  250000
#define INTERVAL1 10
#define INTERVAL2 1
#define INTERVAL3 1

/* Maximum allowed mountpoints and icons
 */
#define MAXICONS   20
#define MAXMPOINTS 20

/* System-wide config file
 */
#define SYSCONFIG "/etc/system.wmmount"

/* Config file defaults
 */
#define MOUNTCMD  "/bin/mount %m"
#define UMOUNTCMD "/bin/umount %m"
#define OPENCMD   "/usr/bin/x-terminal-emulator -T '%n - %m' -e mc %m"
#define NAMEFONT  "-*-helvetica-bold-r-*-*-10-*-*-*-*-*-*-*"
#define USAGEFONT "-*-helvetica-medium-r-*-*-10-*-*-*-*-*-*-*"


/* End of user defines
 *
 * -- AVOID MAKING MODIFICATIONS BEYOND THIS POINT --
 */


/* Includes
 */

#include "dockapp.h"

#include <sys/stat.h>
#include <sys/wait.h>
#include <signal.h>
#ifdef __linux__
#include <mntent.h>
#include <sys/vfs.h>
#endif
#ifdef __FreeBSD__
#include <sys/param.h>
#include <sys/mount.h>
#endif

#include "xpm/main.xpm"


/* Variables
 */

Pixmap pm_main;
Pixmap pm_main_mask;

char *mountcmd=NULL;
char *umountcmd=NULL;
char *opencmd=NULL;
char *namefont=NULL;
char *usagefont=NULL;

Font f_name, f_usage;

Pixmap pm_icon[MAXICONS];
int icons=0;

struct mpinfo {
  char *mpoint;
  char *name;
  char *mountcmd;
  char *umountcmd;
  char *opencmd;
  int icon;
  int udisplay;
} mpoint[MAXMPOINTS];
int mpoints=0;

int namex, namey, usagex, usagey;

#define BUTTON1 (int)(1<<0)
#define BUTTON2 (int)(1<<1)
#define BUTTON3 (int)(1<<2)
#define DCLICK  (int)(1<<3)
#define PREV    BUTTON1
#define NEXT    BUTTON2

int state=0;

int select1=0;

int counter1=0;
int counter2=0;
int counter3=0;

bool mounted;
int usage;
char usagestr[8];


/* Functions and procedures
 */

int main(int argc, char **argv);

void init(int argc, char * const *argv);
void done();

void mainloop();

void buttonpress(XButtonEvent *xev);
void buttonrelease(XButtonEvent *xev);

int config_init();
void config_done();

void xpm_init();
void xpm_done();

void font_init();
void font_done();

void checkmount(bool forced);

void drawall();
void drawbtns(int btnmask);
void drawbtn(int x, int y, int w, int h, bool down);

void execcmd(char *cmd, char *path);

/* Implementation
 */

int main(int argc, char **argv) {
  /* Initialize
   */
  init(argc, (char * const *)argv);

  /* Initialize dockapp
   */
  dockapp_init(argc, (char * const *)argv);

  /* Open dockapp
   */
  if(!dockapp_open()) {
    
    /* Read config file
     */
    if(!config_init()) {

      /* Initialize xpms
       */
      xpm_init();
      
      /* Initialize fonts
       */
      font_init();

      /* Check mounts
       */
      checkmount(true);

      /* Draw everything
       */
      drawall();
      
      /* Execute main loop
       */
      mainloop();

      /* Free fonts
       */
      font_done();

      /* Free xpms
       */
      xpm_done();

      /* Free config file data
       */
      config_done();

    }

    /* Close dockapp
     */
    dockapp_close();
    
  }

  /* Free dockapp
   */
  dockapp_done();

  /* Free
   */
  done();

  /* Finish
   */
  return 0;
}

void init(int argc, char * const *argv) {
  /* Process command-line arguments
   */
  int option_index;
  int option_c;
  struct option long_options[] = {
    {"help", 0, NULL, 'h' },
    {NULL,   0, NULL, '\0'}
  };
  while(true) {
    option_index=0;
    option_c=getopt_long_only(argc, (char * const *)argv, "-h", long_options, &option_index);
    if(option_c==-1)
      break;
    if(option_c==1) {
      fprintf(stderr, "%s: unwanted argument -- '%s'\n", argv[0], optarg);
    }
    if(option_c>1) {
      switch(option_c) {
      case 'h':
	printf("wmmount - The Window Maker universal mount point\n");
        printf("\n");
	printf("17/06/99  Release 1.0 beta2\n");
        printf("Copyright (C) 1999  Sam Hawker <shawkie@geocities.com>\n");
        printf("This software comes with ABSOLUTELY NO WARRANTY\n");
        printf("This software is free software, and you are welcome to redistribute it\n");
        printf("under certain conditions\n");
	printf("See the README file for details.\n");
	printf("\n");
        printf("usage:\n");
        printf("\n");
	printf("   %s [options] [-- dockapp_options]\n", argv[0]);
        printf("\n");
        printf("options:\n");
        printf("\n");
	printf("   -h, --help           Show this help message\n");
        printf("\n");
        printf("dockapp options:\n");
        printf("\n");
        printf("   -h, --help           Show dockapp help message\n");
	exit(0);
      }
    }
  }
}

void done() {
  /* Free command-line option arguments
   */
}

void mainloop() {
  /* Do the main loop
   */
  XEvent xev;

  fd_set fds;
  struct timeval tv;

  XSelectInput(dockapp_d, dockapp_w_active, ButtonPressMask | ButtonReleaseMask | ExposureMask);
  XMapWindow(dockapp_d, dockapp_w_main);

  while(true) {
    while(XPending(dockapp_d)) {
      XNextEvent(dockapp_d, &xev);
      switch(xev.type) {
      case Expose:
	dockapp_expose();
	break;
      case ButtonPress:
        buttonpress(&xev.xbutton);
	break;
      case ButtonRelease:
	buttonrelease(&xev.xbutton);
	break;
      case ClientMessage:
	if(xev.xclient.data.l[0]==dockapp_a_delwin)
          return;
      }
    }
    FD_ZERO(&fds);
    FD_SET(ConnectionNumber(dockapp_d), &fds);
    tv.tv_sec=(long)INTERVAL/1000000L;
    tv.tv_usec=(long)INTERVAL%1000000L;
    if(!select(ConnectionNumber(dockapp_d)+1, &fds, NULL, NULL, &tv)) {
      counter1++;
      if(counter1>=INTERVAL1) {
	counter1=0;
	/* Update interval
	 */
	checkmount(false);
      }
      counter2++;
      if(counter2>=INTERVAL2) {
	counter2=0;
	/* Button repeat interval
	 */
	if(state & (PREV | NEXT)) {
	  if(state & PREV)
	    select1--;
	  else
	    select1++;
	  if(select1<0)
	    select1=mpoints-1;
	  if(select1>=mpoints)
	    select1=0;
	  checkmount(true);
	}
      }
      counter3++;
      if(counter3>=INTERVAL3) {
	counter3=0;
	/* Double-click interval/timeout
	 */
	state&=~DCLICK;
      }
    }
  }
}

void buttonpress(XButtonEvent *xev) {
  /* Process button press events
   */
  int x=xev->x-(dockapp_size/2-32);
  int y=xev->y-(dockapp_size/2-32);

  char *cmd;

  if(x>=33 && y>=47 && x<=45 && y<=57) {
    select1--;
    if(select1<0)
      select1=mpoints-1;
    counter2=0;
    state|=PREV;
    checkmount(true);
    return;
  }

  if(x>=46 && y>=47 && x<=58 && y<=57) {
    select1++;
    if(select1>mpoints-1)
      select1=0;
    counter2=0;
    state|=NEXT;
    checkmount(true);
    return;
  }

  if(x>=5 && y>=47 && x<=32 && y<=57) {
    if(mounted) {
      cmd=mpoint[select1].umountcmd;
      if(cmd==NULL)
	cmd=umountcmd;
      if(cmd==NULL)
	cmd=UMOUNTCMD;
    }
    else {
      cmd=mpoint[select1].mountcmd;
      if(cmd==NULL)
	cmd=mountcmd;
      if(cmd==NULL)
	cmd=MOUNTCMD;
    }
    execcmd(cmd, NULL);

    return;
  }

  if(x>=4 && y>=4 && x<=59 && y<=42) {
    if(mounted) {
      if(state & DCLICK) {

	cmd=mpoint[select1].opencmd;
	if(cmd==NULL)
	  cmd=opencmd;
        if(cmd==NULL)
	  cmd=OPENCMD;
        execcmd(cmd, NULL);

	state&=~DCLICK;
      }
      else {
	state|=DCLICK;
	counter3=0;
      }
    }
    return;
  }

}

void buttonrelease(XButtonEvent *xev) {
  /* Process button release events
   */
  state&=~(PREV | NEXT);
  drawbtns(PREV | NEXT);
  dockapp_expose();
}

int config_init() {
  /* Process config file
   */

  char *home;
  char *filename;

  FILE *f;

  char buf[256];

  int keynum1;
  char *keyarg1;
  static char *keys1[] = {
    "mountcmd=",
    "umountcmd=",
    "opencmd=",
    "namefont=",
    "usagefont=",
    "icon ",
    "mountpoint ",
    ""
  };
  
  int keynum2;
  char *keyarg2;
  static char *keys2[] = {
    "mountpoint ",
    "name=",
    "iconnumber=",
    "usagedisplay=",
    "mountcmd=",
    "umountcmd=",
    "opencmd=",
    ""
  };

  home=getenv("HOME");
  filename=(char *)malloc(strlen(home)+strlen("/.wmmount")+1);
  sprintf(filename, "%s/.wmmount", home);
  f=fopen(filename, "r");
  free(filename);
  if(f==NULL) {
/*  fprintf(stderr, "%s: User config file not found\n", dockapp_argv[0]); */
    filename=SYSCONFIG;
    f=fopen(filename, "r");
    if(f==NULL) {
      fprintf(stderr, "%s: System config file not found\n", dockapp_argv[0]);
      return -1;
    }
  }

  while(1) {
    fgets(buf, 256, f);
    if(feof(f)!=0)
      break;
    buf[strlen(buf)-1]='\0';

    keynum1=0;
    while(1) {
      if(strncmp(buf, keys1[keynum1], strlen(keys1[keynum1]))==0)
	break;
      keynum1++;
    }

    keyarg1=buf+strlen(keys1[keynum1]);

    if(keynum1==0)
      mountcmd=strdup(keyarg1);
    if(keynum1==1)
      umountcmd=strdup(keyarg1);
    if(keynum1==2)
      opencmd=strdup(keyarg1);
    if(keynum1==3)
      namefont=strdup(keyarg1);
    if(keynum1==4)
      usagefont=strdup(keyarg1);
    if(keynum1==5) {
      if(icons<MAXICONS) {
	dockapp_createpixmap(NULL, NULL, keyarg1, &pm_icon[icons], NULL, NULL, NULL);
	icons++;
      }
    }
    if(keynum1==6) {
      while(mpoints<MAXMPOINTS && feof(f)==0) {
	mpoint[mpoints].mountcmd=NULL;
	mpoint[mpoints].umountcmd=NULL;
	mpoint[mpoints].opencmd=NULL;
	mpoint[mpoints].mpoint=strdup(keyarg1);

	while(1) {
	  fgets(buf, 256, f);
	  if(feof(f)!=0)
	    break;
	  buf[strlen(buf)-1]='\0';

	  keynum2=0;
	  while(1) {
	    if(strncmp(buf, keys2[keynum2], strlen(keys2[keynum2]))==0)
	      break;
	    keynum2++;
	  }

	  keyarg2=buf+strlen(keys2[keynum2]);

	  if(keynum2==0)
	    break;
	  if(keynum2==1)
	    mpoint[mpoints].name=strdup(keyarg2);
	  if(keynum2==2)
	    sscanf(keyarg2, "%i", &mpoint[mpoints].icon);
	  if(keynum2==3)
	    sscanf(keyarg2, "%i", &mpoint[mpoints].udisplay);
	  if(keynum2==4)
	    mpoint[mpoints].mountcmd=strdup(keyarg2);
	  if(keynum2==5)
	    mpoint[mpoints].umountcmd=strdup(keyarg2);
	  if(keynum2==6)
	    mpoint[mpoints].opencmd=strdup(keyarg2);
	}
	mpoints++;
      }
    }

  }

  fclose(f);

  if(mpoints==0) {
    fprintf(stderr, "%s: No mountpoints specified\n", dockapp_argv[0]);
    return -1;
  }

  return 0;
}

void config_done() {
  /* Free config file stuff
   */
  free(mountcmd);
  free(umountcmd);
  free(opencmd);
  free(namefont);
  free(usagefont);

  while(mpoints>0) {
    mpoints--;
    free(mpoint[mpoints].mpoint);
    free(mpoint[mpoints].name);
    free(mpoint[mpoints].mountcmd);
    free(mpoint[mpoints].umountcmd);
    free(mpoint[mpoints].opencmd);
  }
}

void xpm_init() {
  /* Create pixmaps from xpm data
   */
  dockapp_createpixmap(main_xpm, NULL, NULL, &pm_main, &pm_main_mask, NULL, NULL);

  XCopyArea(dockapp_d, pm_main_mask, dockapp_pm_mask, dockapp_gc_mask, 0, 0, 64, 64, 0, 0);
  dockapp_installmask();
}

void xpm_done() {
  /* Free xpms
   */

  /* Done automatically by X, so I won't bother
   */
}

void font_init() {
  /* Initialize fonts
   */
  XFontStruct *xfs;
  char *font;

  font=namefont;
  if(font==NULL)
    font=NAMEFONT;
  xfs=XLoadQueryFont(dockapp_d, font);
  if(xfs==NULL) {
    fprintf(stderr, "wmmount: font %s not found\n", font);
    exit(1);
  }

  f_name=xfs->fid;
  namex=6-xfs->min_bounds.lbearing;
  namey=5+xfs->ascent;
  usagey=namey;
  usagey+=xfs->descent;
  XFreeFontInfo(NULL, xfs, 1);

  font=usagefont;
  if(font==NULL)
    font=USAGEFONT;
  xfs=XLoadQueryFont(dockapp_d, USAGEFONT);
  if(xfs==NULL) {
    fprintf(stderr, "wmmount: font %s not found\n", font);
    exit(1);
  }
  
  f_usage=xfs->fid;
  usagex=6-xfs->min_bounds.lbearing;
  usagey+=xfs->ascent;
  XFreeFontInfo(NULL, xfs, 1);
}

void font_done() {
  /* Free fonts
   */

  /* Done automatically by X
   */
}

void checkmount(bool forced) {
  /* Check mount and usage
   */

  bool getmounted;
  int getusage, getavail;

  dev_t dev1, dev2;
  struct stat st;
  struct statfs sfs;
  char *buf;

  float value;
  char unit;

  getmounted=false;
  if(strcmp(mpoint[select1].mpoint, "/")==0)
    /* I think I can assume / is mounted.
     */
    getmounted=true;
  else {
    /* Check that the directory above the mount point is on a different device
     */
    buf=strdup(mpoint[select1].mpoint);

    stat(buf, &st);
    dev1=st.st_dev;

    if(buf[strlen(buf)-1]=='/')
      buf[strlen(buf)-1]='\0';
    *(strrchr(buf, '/')+1)='\0';

    stat(buf, &st);
    dev2=st.st_dev;

    if(memcmp(&dev1, &dev2, sizeof(dev_t))!=0)
      getmounted=true;

    free(buf);
  }

  if(getmounted && mpoint[select1].udisplay!=0) {
    statfs(mpoint[select1].mpoint, &sfs);
    getusage=(sfs.f_blocks-sfs.f_bfree);
    getavail=(sfs.f_bavail);
  }

  if(getmounted && (!mounted || (getusage!=usage) || forced)) {

    value=getusage;
    switch(mpoint[select1].udisplay) {
    case 1:
      value=getavail;
    case 2:
      value*=(float)sfs.f_bsize;

      value/=1024.0;
      unit='k';
      if(value>=999.5) {
	value/=1024.0;
	unit='M';
      }
      if(value>=999.5) {
	value/=1024.0;
	unit='G';
      }

      if(value<9.995) {
	sprintf(usagestr, "%1.02f%cB", value, unit);
	break;
      }
      if(value<99.95) {
	sprintf(usagestr, "%1.01f%cB", value, unit);
	break;
      }
      if(value<999.5) {
	sprintf(usagestr, "%1.0f%cB", value, unit);
	break;
      }
      strcpy(usagestr, "E big");
      break;
    case 3:
      if(getusage+getavail==0)
	strcpy(usagestr, "E div0");
      else
	sprintf(usagestr, "%.01f%%", getusage*100.0/(getusage+getavail));
      break;
    default:
      strcpy(usagestr, "");
    }
  }

  if((getmounted!=mounted) || (getmounted && getusage!=usage && mpoint[select1].udisplay!=0) || forced) {
    mounted=getmounted;
    usage=getusage;
    drawall();
    dockapp_expose();
  }

}

void drawall() {
  /* Draw everything
   */
  XCopyArea(dockapp_d, pm_main, dockapp_pm_main, dockapp_gc_clip, 0, 0, 64, 64, 0, 0);

  XCopyArea(dockapp_d, pm_icon[mpoint[select1].icon], dockapp_pm_main, dockapp_gc_main, 0, 0, 32, 10, 25, 30);

  XSetFont(dockapp_d, dockapp_gc_main, f_name);
  XSetForeground(dockapp_d, dockapp_gc_main, dockapp_color0);
  XDrawString(dockapp_d, dockapp_pm_main, dockapp_gc_main, namex, namey, (char *)mpoint[select1].name, strlen(mpoint[select1].name));

  if(mounted) {
    XSetFont(dockapp_d, dockapp_gc_main, f_usage);
    XSetForeground(dockapp_d, dockapp_gc_main, dockapp_color0);
    XDrawString(dockapp_d, dockapp_pm_main, dockapp_gc_main, usagex, usagey, (char *)usagestr, strlen(usagestr));
  }

  drawbtns(BUTTON1 | BUTTON2 | BUTTON3);
}

void drawbtns(int btnmask) {
  /* Draw buttons
   */
  if(btnmask & BUTTON1)
    drawbtn(33, 47, 13, 11, (state & BUTTON1));
  if(btnmask & BUTTON2)
    drawbtn(46, 47, 13, 11, (state & BUTTON2));
  if(btnmask & BUTTON3)
    drawbtn(5, 47, 28, 11, mounted);
}

void drawbtn(int x, int y, int w, int h, bool down) {
  /* Draw single button
   */
  if(down) {
    XCopyArea(dockapp_d, pm_main, dockapp_pm_main, dockapp_gc_main, x, y, 1, h-1, x+w-1, y+1);
    XCopyArea(dockapp_d, pm_main, dockapp_pm_main, dockapp_gc_main, x+w-1, y+1, 1, h-1, x, y);
    XCopyArea(dockapp_d, pm_main, dockapp_pm_main, dockapp_gc_main, x, y, w-1, 1, x+1, y+h-1);
    XCopyArea(dockapp_d, pm_main, dockapp_pm_main, dockapp_gc_main, x+1, y+h-1, w-1, 1, x, y);
  }
  else
    XCopyArea(dockapp_d, pm_main, dockapp_pm_main, dockapp_gc_main, x, y, w, h, x, y);
}

void execcmd(char *cmd, char *path) {
  /* Execute external command (without using /bin/sh)
   */

  /* This has got quite complicated now
   *
   * It splits the command-line into its arguments, performs substitutions and
   * then executes the command
   *
   * It doesn't wait for a return and children are reaped using SIGCHLD
   */

  int argc=1;
  char **argv=NULL;

  int i, j;
  char c;

  int buflen=strlen(cmd)+1;
  char *buf=(char *)malloc(buflen);

  bool quoted=false;
  char quotec;

  char *subst;

  int pid;

  signal(SIGCHLD, SIG_IGN);

  j=0;
  for(i=0;i<strlen(cmd);i++) {
    c=cmd[i];
    if(!quoted && (c=='\'' || c=='\"')) {
      quoted=true;
      quotec=c;
      continue;
    }
    if(quoted && c==quotec) {
      quoted=false;
      continue;
    }
    if(c=='\\') {
      i++;
      if(i<strlen(cmd))
	c=cmd[i];
      else
	break;
    }
    if(c=='%') {
      i++;
      if(i<strlen(cmd))
	c=cmd[i];
      else
	break;
      subst=NULL;
      if(c=='m')
	subst=mpoint[select1].mpoint;
      if(c=='n')
	subst=mpoint[select1].name;
      if(subst!=NULL) {
	buflen+=strlen(subst)-2;
	buf=(char *)realloc(buf, buflen);
	strcpy(buf+j, subst);
	j+=strlen(subst);
	continue;
      }
    }
    if(!quoted && c==' ') {
      c='\0';
      argc++;
    }
    buf[j]=c;
    j++;
  }
  buf[j]='\0';

  argv=(char **)malloc((argc+1)*sizeof(char *));

  j=0;
  for(i=0;i<argc;i++) {
    argv[i]=buf+j;
    j+=strlen(buf+j)+1;
  }
  argv[argc]=NULL;

  pid=fork();
  if(pid==0) {
    if(path!=NULL)
      chdir(path);
    if(execvp(argv[0], argv) == -1) {
      perror(argv[0]);
    }
    free(argv);
    free(buf);
    exit(127);
  }
}
